File: 42_topological_sort/42_7_supersequence.py
def construct_graph(arr: list[str])->dict[str, list[str]]:
    graph = {}
    for s in arr:
        n = len(s)
        for i in range(n):
            if s[i] not in graph:
                graph[s[i]] = []

            if i < n - 1:
                graph[s[i]].append(s[i+1])

    return graph

def topological_sort(graph: dict[str, list[str]])->list[str]:
    in_degrees = {}
    for key in graph.keys():
        in_degrees[key] = 0
    
    for node, nbrs in graph.items():
        for nbr in nbrs:
            in_degrees[nbr] += 1
    
    zero_degrees = []
    for node, in_degree in in_degrees.items():
        if in_degree == 0:
            zero_degrees.append(node)

    topo_order = []
    while len(zero_degrees) > 0:
        node = zero_degrees.pop()
        topo_order.append(node)

        for nbr in graph[node]:
            in_degrees[nbr] -= 1
            if in_degrees[nbr] == 0:
                zero_degrees.append(nbr)
    
    return topo_order


def supersequence(arr: list[str])->bool:
    graph = construct_graph(arr)
    topo_order = topological_sort(graph)
    return len(topo_order) == len(graph)
    

if __name__ == "__main__":
    # Example1
    arr = ["abc", "bde", "df", "cfe"]
    print(supersequence(arr))

    # Example2
    arr = ["ab", "ba"]
    print(supersequence(arr))
    
    # Example3
    arr = ["aa"]
    print(supersequence(arr))

File: 42_topological_sort/42_7_supersequence.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>

std::unordered_map<char, std::vector<char>> construct_graph(const std::vector<std::string>& arr) {
    std::unordered_map<char, std::vector<char>> graph;

    for (const std::string& s : arr) {
        int n = static_cast<int>(s.size());
        for (int i = 0; i < n; i++) {
            if (graph.find(s[i]) == graph.end()) {
                graph[s[i]] = {};
            } 
            
            if (i < n - 1) {
                graph[s[i]].push_back(s[i+1]);
            }
        }
    }

    return graph;
}

std::vector<char> topological_sort(const std::unordered_map<char, std::vector<char>>& graph) {
    std::unordered_map<char, int> in_degrees;
    for (const auto& [node, _] : graph) {
        in_degrees[node] = 0;
    }

    for (const auto& [node, nbrs] : graph) {
        for (char nbr : nbrs) {
            in_degrees[nbr]++;
        }
    }

    std::vector<char> zero_degrees;
    for (const auto& [node, in_degree] : in_degrees) {
        if (in_degree == 0) {
            zero_degrees.push_back(node);
        }
    }

    std::vector<char> topo_order;
    while (!zero_degrees.empty()) {
        char node = zero_degrees.back();
        zero_degrees.pop_back();
        topo_order.push_back(node);

        for (char nbr : graph.at(node)) {
            in_degrees[nbr]--;
            if (in_degrees[nbr] == 0) {
                zero_degrees.push_back(nbr);
            }
        }
    }

    return topo_order;
}

bool supersequence(const std::vector<std::string>& arr) {
    std::unordered_map<char, std::vector<char>> graph = construct_graph(arr);
    std::vector<char> topo_order = topological_sort(graph);
    return topo_order.size() == graph.size();
}

int main() {
    // Example1
    std::vector<std::string> arr = {"abc", "bde", "df", "cfe"};
    std::cout << (supersequence(arr) ? "true" : "false") << std::endl;

    // Example2
    arr = {"ab", "ba"};
    std::cout << (supersequence(arr) ? "true" : "false") << std::endl;

    // Example3
    arr = {"aa"};
    std::cout << (supersequence(arr) ? "true" : "false") << std::endl;

    return 0;
}